src/backend/dx12/src/conv.rs

```diff
@@ -486,54 +486,47 @@ pub fn map_image_resource_state(
     layout: image::Layout,
 ) -> D3D12_RESOURCE_STATES {
     use self::image::Access;
-    // `D3D12_RESOURCE_STATE_PRESENT` is the same as COMMON (general state)
-    if layout == image::Layout::Present {
-        return D3D12_RESOURCE_STATE_PRESENT;
-    }
-
-    // Mutable states
-    if access.contains(Access::SHADER_WRITE) {
-        return D3D12_RESOURCE_STATE_UNORDERED_ACCESS;
-    }
-    if access.contains(Access::DEPTH_STENCIL_ATTACHMENT_WRITE) {
-        return D3D12_RESOURCE_STATE_DEPTH_WRITE;
-    }
-    if access.contains(Access::COLOR_ATTACHMENT_READ)
-        || access.contains(Access::COLOR_ATTACHMENT_WRITE)
-    {
-        return D3D12_RESOURCE_STATE_RENDER_TARGET;
-    }
-
-    // `TRANSFER_WRITE` requires special handling as it requires RESOLVE_DEST | COPY_DEST
-    // but only 1 write-only allowed. We do the required translation before the commands.
-    // We currently assume that `COPY_DEST` is more common state than out of renderpass resolves.
-    // Resolve operations need to insert a barrier before and after the command to transition from and
-    // into `COPY_DEST` to have a consistent state for srcAccess.
-    if access.contains(Access::TRANSFER_WRITE) {
-        return D3D12_RESOURCE_STATE_COPY_DEST;
-    }
 
-    // Read-only states
-    let mut state = D3D12_RESOURCE_STATE_COMMON;
+    match layout {
+        // the same as COMMON (general state)
+        image::Layout::Present => D3D12_RESOURCE_STATE_PRESENT,
+        image::Layout::ColorAttachmentOptimal => D3D12_RESOURCE_STATE_RENDER_TARGET,
+        image::Layout::DepthStencilAttachmentOptimal => D3D12_RESOURCE_STATE_DEPTH_WRITE,
+        // `TRANSFER_WRITE` requires special handling as it requires RESOLVE_DEST | COPY_DEST
+        // but only 1 write-only allowed. We do the required translation before the commands.
+        // We currently assume that `COPY_DEST` is more common state than out of renderpass resolves.
+        // Resolve operations need to insert a barrier before and after the command to transition
+        // from and into `COPY_DEST` to have a consistent state for srcAccess.
+        image::Layout::TransferDstOptimal => D3D12_RESOURCE_STATE_COPY_DEST,
+        image::Layout::TransferSrcOptimal => D3D12_RESOURCE_STATE_COPY_SOURCE,
+        // At this point, we assume that the mutable access flags have already been covered by the
+        // layout checks, and we ignore them.
+        image::Layout::General if access.contains(Access::SHADER_WRITE) => D3D12_RESOURCE_STATE_UNORDERED_ACCESS,
+        image::Layout::ShaderReadOnlyOptimal | // could be moved outside
+        image::Layout::DepthStencilReadOnlyOptimal |
+        image::Layout::General |
+        image::Layout::Undefined |
+        image::Layout::Preinitialized => {
+            // Read-only states
+            let mut state = D3D12_RESOURCE_STATE_COMMON;
+
+            if access.contains(Access::TRANSFER_READ) {
+                state |= D3D12_RESOURCE_STATE_COPY_SOURCE;
```

Comment:
as resource states are not flags we should just return them

Reply:
no, I think you are mistaken here:
> The resource state usage bits that are used to describe a resource state are divided into read-only and read/write states. The reference topic for the D3D12_RESOURCE_STATES indicates the read/write access level for each bit in the enumeration.
> At most, only one read/write bit can be set for any resource. If a write bit is set, then no read-only bit may be set for that resource. If no write bit is set, then any number of read bits may be set.

Reply:
oh right, mixed it up thanks.